File: src/main/java/rars/venus/settings/editor/controllers/SyntaxStyleSettingsController.java
package rars.venus.settings.editor.controllers;

import org.jetbrains.annotations.NotNull;
import rars.settings.TokenSettingKey;
import rars.venus.editors.TextEditingArea;
import rars.venus.editors.TokenStyle;
import rars.venus.settings.editor.views.SyntaxStyleView;

import javax.swing.event.ChangeListener;
import java.util.List;

public final class SyntaxStyleSettingsController {
    private final @NotNull SyntaxStyleView view;
    private final @NotNull EditorSettingsController parentController;
    private final @NotNull TextEditingArea textArea;
    private @NotNull TokenSettingKey currentKey;

    public SyntaxStyleSettingsController(
        final @NotNull SyntaxStyleView view,
        final @NotNull EditorSettingsController parentController,
        final @NotNull TextEditingArea textArea
    ) {
        this.view = view;
        this.parentController = parentController;
        this.textArea = textArea;
        this.currentKey = TokenSettingKey.COMMENT; // initial value
        initializeView();

        // initialize selection based on the list
        final var selected = this.view.tokenTypesList.getSelectedValue();
        if (selected != null) {
            setCurrentKey(selected);
        } else {
            setCurrentKey(this.currentKey);
        }
    }

    private void initializeView() {
        final ChangeListener tokenStyleChangeListener = e -> {
            final var tokenStyle = getTokenStyleFromView();
            this.parentController.settingsTheme.tokenStyles.put(this.currentKey, tokenStyle);
            TokenSettingKey
                .getTokenTypesForSetting(this.currentKey)
                .forEach(key -> this.textArea
                    .setTokenStyle(key, tokenStyle)
                );
        };
        List.of(
            view.useForeground,
            view.foregroundColorButton,
            view.useBackground,
            view.backgroundColorButton,
            view.isBold,
            view.isItalic,
            view.isUnderline
        ).forEach(component -> component.addChangeListener(tokenStyleChangeListener));

        // selection listener for the token types list
        view.tokenTypesList.addListSelectionListener(e -> {
            if (!e.getValueIsAdjusting()) {
                final var key = view.tokenTypesList.getSelectedValue();
                if (key != null) {
                    setCurrentKey(key);
                }
            }
        });
    }

    private @NotNull TokenStyle getTokenStyleFromView() {
        final var foreground = this.view.useForeground.isSelected() ? this.view.foregroundColorButton.getColor() : null;
        final var background = this.view.useBackground.isSelected() ? this.view.backgroundColorButton.getColor() : null;
        final var isBold = this.view.isBold.isSelected();
        final var isItalic = this.view.isItalic.isSelected();
        final var isUnderline = this.view.isUnderline.isSelected();
        return new TokenStyle(foreground, background, isBold, isItalic, isUnderline);
    }

    public void setCurrentKey(final @NotNull TokenSettingKey key) {
        this.currentKey = key;
        this.setView(key);
    }

    public void resetButtonValues() {
        this.setView(this.currentKey);
    }

    private void setView(final @NotNull TokenSettingKey key) {
        final var settingsTheme = this.parentController.settingsTheme;
        final var tokenStyle = settingsTheme.tokenStyles.get(key);
        final var foreground = tokenStyle.foreground();
        if (foreground != null) {
            this.view.useForeground.setSelected(true);
            this.view.foregroundColorButton.setColor(foreground);
        } else {
            this.view.useForeground.setSelected(false);
            this.view.foregroundColorButton.setColor(settingsTheme.foregroundColor);
        }
        final var background = tokenStyle.background();
        if (background != null) {
            this.view.useBackground.setSelected(true);
            this.view.backgroundColorButton.setColor(background);
        } else {
            this.view.useBackground.setSelected(false);
            this.view.backgroundColorButton.setColor(settingsTheme.backgroundColor);
        }
        this.view.isBold.setSelected(tokenStyle.isBold());
        this.view.isItalic.setSelected(tokenStyle.isItalic());
        this.view.isUnderline.setSelected(tokenStyle.isUnderline());
    }
}

File: src/main/java/rars/venus/settings/editor/controllers/SyntaxStyleSettingsController.kt
package rars.venus.settings.editor.controllers

import rars.settings.TokenSettingKey
import rars.settings.TokenSettingKey.Companion.getTokenTypesForSetting
import rars.venus.editors.TextEditingArea
import rars.venus.editors.TokenStyle
import rars.venus.settings.editor.views.SyntaxStyleView
import javax.swing.AbstractButton
import javax.swing.event.ChangeListener

class SyntaxStyleSettingsController(
    private val view: SyntaxStyleView,
    private val parentController: EditorSettingsController,
    private val textArea: TextEditingArea
) {
    private var currentKey: TokenSettingKey

    init {
        this.currentKey = TokenSettingKey.COMMENT // dummy value
        view.initializeView()
    }

    private fun SyntaxStyleView.initializeView() {
        val tokenStyleChangeListener = ChangeListener {
            val tokenStyle = tokenStyleFromView
            parentController.settingsTheme.tokenStyles.put(currentKey, tokenStyle)
            getTokenTypesForSetting(currentKey).forEach { key ->
                textArea.setTokenStyle(key, tokenStyle)
            }
        }
        arrayOf<AbstractButton>(
            useForeground,
            foregroundColorButton,
            useBackground,
            backgroundColorButton,
            isBold,
            isItalic,
            isUnderline
        ).forEach { component ->
            component.addChangeListener(tokenStyleChangeListener)
        }
    }

    private val tokenStyleFromView: TokenStyle
        get() = view.run {
            val foreground = if (useForeground.isSelected) foregroundColorButton.color else null
            val background = if (useBackground.isSelected) backgroundColorButton.color else null
            val isBold = isBold.isSelected
            val isItalic = isItalic.isSelected
            val isUnderline = isUnderline.isSelected
            TokenStyle(foreground, background, isBold, isItalic, isUnderline)
        }

    fun setCurrentKey(key: TokenSettingKey) {
        this.currentKey = key
        view.setView(key)
    }

    fun resetButtonValues() {
        view.setView(this.currentKey)
    }

    private fun SyntaxStyleView.setView(key: TokenSettingKey) {
        val settingsTheme = parentController.settingsTheme
        val tokenStyle = settingsTheme.tokenStyles[key]!!
        val newForeground = tokenStyle.foreground
        if (newForeground != null) {
            useForeground.isSelected = true
            foregroundColorButton.color = newForeground
        } else {
            useForeground.isSelected = false
            foregroundColorButton.color = settingsTheme.foregroundColor
        }
        val background = tokenStyle.background
        if (background != null) {
            useBackground.isSelected = true
            backgroundColorButton.color = background
        } else {
            useBackground.isSelected = false
            backgroundColorButton.color = settingsTheme.backgroundColor
        }
        isBold.isSelected = tokenStyle.isBold
        isItalic.isSelected = tokenStyle.isItalic
        isUnderline.isSelected = tokenStyle.isUnderline
    }
}
